main/util/src/mill/util/Version.scala

```diff
@@ -0,0 +1,161 @@
+    Some(major),
+    minor.map("." + _),
+    micro.map("." + _),
+    qualifierWithSep.map(q => q._2 + q._1)
+  ).flatten.mkString
+
+  def isNewerThan(other: Version)(implicit ordering: Ordering[Version]): Boolean =
+    ordering.compare(this, other) > 0
+
+  def chooseNewest(versions: Version*)(implicit ordering: Ordering[Version]): Version =
+    versions.foldLeft(this)((best, next) =>
+      if (next.isNewerThan(best)) next else best
+    )
+
+  def isAtLeast(other: Version)(implicit ordering: Ordering[Version]): Boolean =
+    ordering.compare(this, other) >= 0
+
+  def asMaven: MavenVersion = new MavenVersion(this)
+  def asOsgiVersion: OsgiVersion = new OsgiVersion(this)
+  def asIgnoreQualifierVersion: IgnoreQualifierVersion = new IgnoreQualifierVersion(this)
+}
+
+final class MavenVersion(val underlying: Version) extends AnyVal {
+  def isNewerThan(other: MavenVersion): Boolean = {
+    underlying.isNewerThan(other.underlying)(Version.MavenOrdering)
+  }
+}
+final class OsgiVersion(val underlying: Version) extends AnyVal {
+  def isNewerThan(other: OsgiVersion): Boolean =
+    underlying.isNewerThan(other.underlying)(Version.OsgiOrdering)
+}
+final class IgnoreQualifierVersion(val underlying: Version) extends AnyVal {
+  def isNewerThan(other: IgnoreQualifierVersion): Boolean =
+    underlying.isNewerThan(other.underlying)(Version.IgnoreQualifierOrdering)
+}
+
+object Version {
+
+  /**
+   * Missing minor or micro versions are equal to zero,
+   * a qualifier is ignored,
+   * stable ordering.
+   */
+  object IgnoreQualifierOrdering extends Ordering[Version] {
+    override def compare(l: Version, r: Version): Int =
+      l.major - r.major match {
+        case 0 =>
+          l.minor.getOrElse(0) - r.minor.getOrElse(0) match {
+            case 0 => l.micro.getOrElse(0) - r.micro.getOrElse(0)
+            case x => x
+          }
+        case x => x
+      }
+  }
+
+  /**
+   * Missing minor or micro versions are equal to zero,
+   * a qualifier is higher that no qualifier,
+   * qualifiers are sorted alphabetically,
+   * stable ordering.
+   */
+  object OsgiOrdering extends Ordering[Version] {
+    override def compare(l: Version, r: Version): Int =
+      l.major - r.major match {
+        case 0 =>
+          l.minor.getOrElse(0) - r.minor.getOrElse(0) match {
+            case 0 =>
+              l.micro.getOrElse(0) - r.micro.getOrElse(0) match {
+                case 0 =>
+                  l.qualifierWithSep.map(_._1).getOrElse("")
+                    .compareTo(r.qualifierWithSep.map(_._1).getOrElse(""))
+                case x => x
+              }
+            case x => x
+          }
+        case x => x
+      }
+  }
+
+  /**
+   * Missing minor or micro versions are equal to zero,
+   * a qualifier is lower that no qualifier,
+   * qualifiers are sorted alphabetically,
+   * stable ordering.
+   *
+   * TODO: Review ordering wrt Maven 3
+   * TODO: also consider a coursier ordering
```

Comment:
I _think_ coursier may just have built-in versioning things we could piggy back on. Or even just use it if it's exposed.